### DISCUSSION TITLE
Struggling to find a suitable way to implement scrolling and updating data in a loop

### DISCUSSION THREAD

**Discussion by EddRobinson** [tui-cs/Terminal.Gui#2439] [Q&A]
I'll try and explain this the best way I can.
I basically have a background thread that every 10ms sends a load of commands through the serial port and then receives a lot of data back which is decoded and displayed in my program.

I have a List that stores the received data and then displays it in labels. But, I'm having an issue finding the best way to have it constantly updating but also have the list in a scroll-able view.

I originally had a FrameView that displayed all the data and that was refreshing every 10ms and working perfectly, but struggled to implement scrolling.

I've recently switched to a ListView, which was near perfect, scrolling was working and everything was displaying, except it only updated when there was mouse movement / keyboard input. Not ideal.

Using the SetSourceAsync Task, I've managed to achieve the constant update of the values needed, but it interferes with the scrolling and constantly moves back to the top of the list (so you can't scroll down).

I'm unsure what else try or have I overlooked a simple way of implementing this?

Many thanks in advanced

**Comment by BDisp** [comment]
You must update inside the Application.Mainloop.Invoke to use the main thread.

**Reply by EddRobinson** [reply]
I am, that's how I managed to get the data to constantly update as required.

The issue is then, that it interferes with the scrolling

**Comment by BDisp** [comment]
Ok. If you are using a `ListView` add new items to the source and use the `MoveEnd` method to scroll to end. You must not update the `ListView` source but add news items to the List. Don't forget to use `SetNeedsDisplay` method.

**Reply by EddRobinson** [reply]
Hmm... okay. Done some more work on trying to fix my views as I had some left over stuff that may have been affecting it (I'm still learning xD)

So, I have an object list (a) that holds the data ready for my decode methods inside the `MainLoop`.

Before the `MainLoop`, I have a string List (b) that gets populated from 2 parts of the object list (a) mentioned above.
I then initialize the `ListView `& use the string list (b) as it's source.

Inside the `MainLoop `I get the data as needed, decode it and store it in part of object list (a) which is one of the 2 values that needs to be displayed through the string list (b) using the [ListView](url).

I then call `SetNeedsDisplay` ( inside of the `MainLoop`) on the string list (b) but, not managed to make it re-display / update with the new data.

Have I missed / misunderstood something?

p.s. I really appreciate the help

**Reply by BDisp** [reply]
Are you doing this inside the `MainLoop`. Pay attention that the listview must be already instantiated with the source and added to the sujperview, before use it in the mainloop.

```cs
Application.Mainloop.Invoke(() => {
  myListSource.Add("something more");
  listview.MoveEnd();
  //listview.SetNeedsDisplay(); // maybe not needed because the prior method already call it.
});
```

**Reply by EddRobinson** [reply]
```
List<string> listViewSource = new();

            foreach (var data in addressObjects)
            {
                listViewSource.Add($"{data.AddressName}:{data.RecievedData}");
            }

            ListView addressListView = new(listViewSource)
            {
                Width = 150,
                Height = Program.CyclicFrame.Height,
                X = 0,
                Y = 0,
            };
            Program.CyclicFrame.Add(addressListView);

            while (serialDevice != null)
            {
                try
                {
                    Application.MainLoop.Invoke(async () =>
                    {

                        // Loop through the registers that have been selected & Decode the returned data (Which  then gets stored in addressObjects.RecievedData)
                        foreach (var item in selectedRegisters)...

                        addressListView.SetNeedsDisplay();
                    });
                }
                catch
                {
                    // Error Stuff 
                }
            }
```

ListView is already instantiated with the source, and added to the view. But it doesn't seem to update the display in the MainLoop

If I try and add stuff to the list source inside of the MainLoop, nothing gets displayed.

**Reply by BDisp** [reply]
After you add the `ListView` into the `Program.CyclicFrame.Add(addressListView);` and run your app, you must put the `while` statement inside a thread. It must be write before the `Application.Run`.

```cs
Task.Run (() => {
            while (serialDevice != null)
            {
                try
                {
                    Application.MainLoop.Invoke(async () =>
                    {

                        // Loop through the registers that have been selected & Decode the returned data (Which  then gets stored in addressObjects.RecievedData)
                        foreach (var item in selectedRegisters) {
                          listViewSource.Add (item);
                        };
                        
                        addressListView.MoveEnd();
                        //addressListView.SetNeedsDisplay(); // MoveEnd already call SetNeedsDisplay(). uncomment if neeeded
                    });
                }
                catch
                {
                    // Error Stuff 
                }
            }
});
```

**Reply by EddRobinson** [reply]
All of this is already called from a thread using:

`                CyclicThread = new(() => Communications.CyclicCommunications(master, serialDevice, selectedRegisters)) { };
                CyclicThread.Start();`

**Reply by BDisp** [reply]
You must to add the `ListView` with the source into the initial window and not in the thread.

**Reply by EddRobinson** [reply]
Awesome dude, thank you very much.

Is there a way for automatic Text Wrapping using ListView?

**Reply by BDisp** [reply]
No, you have to implement yourself by using the `TextFormatter.WordWrap`, using a custom class with an `override ToString()`. Also by subclassing the `IListDataSource` in the `ListView`. Currently `ListView` has the limitation to have a height length of one. Another option is using the `TextView` with word wrap enabled.

**Reply by tznind** [reply]
In the past I have struggled to get TextView working nicely with scrolling text,  see this PR that I abandonned if you want to go down this route and get a headstart: https://github.com/gui-cs/Terminal.Gui/pull/1838

One way to handle the multi line issue is to render on one line (e.g. with ListView or TableView) and then have a key to pop out the full message into a modal dialog.

There was a cool application written recently by @aaronhudon for lumber packing which had scrolling data using `TableView`.  If your data has timestamps or warning levels or other stuff you want to store in a tabular format then it could be nice:

https://github.com/gui-cs/Terminal.Gui/discussions/2387

Also since it sounds like you are getting into creating a project in Terminal.Gui I thought I would give a ~shout out~ shameless plug to the designer (incase you didn't know about it): https://github.com/gui-cs/TerminalGuiDesigner  .  It speeds up building dialogs and popups and stuff quite a bit and can give you a headstart on making cool apps.

I have seen several users wanting a solution for scrolling readonly text UI.  Maybe we should put the time into developing a dedicated view for it.  Either by subclassing TextView or just from scratch.

**Reply by EddRobinson** [reply]
> No, you have to implement yourself by using the `TextFormatter.WordWrap`, using a custom class with an `override ToString()`. Also by subclassing the `IListDataSource` in the `ListView`. Currently `ListView` has the limitation to have a height length of one. Another option is using the `TextView` with word wrap enabled.

It's fine, I have implemented a workaround that works for my use case using ListView, so I'm happy.

> In the past I have struggled to get TextView working nicely with scrolling text, see this PR that I abandonned if you want to go down this route and get a headstart: #1838
> 
> One way to handle the multi line issue is to render on one line (e.g. with ListView or TableView) and then have a key to pop out the full message into a modal dialog.
> 
> There was a cool application written recently by @aaronhudon for lumber packing which had scrolling data using `TableView`. If your data has timestamps or warning levels or other stuff you want to store in a tabular format then it could be nice:
> 
> #2387
> 
> Also since it sounds like you are getting into creating a project in Terminal.Gui I thought I would give a ~shout out~ shameless plug to the designer (incase you didn't know about it): https://github.com/gui-cs/TerminalGuiDesigner . It speeds up building dialogs and popups and stuff quite a bit and can give you a headstart on making cool apps.
> 
> I have seen several users wanting a solution for scrolling readonly text UI. Maybe we should put the time into developing a dedicated view for it. Either by subclassing TextView or just from scratch.

Thankyou for the useful links. Scrolling readonly text UI would have been perfect for me, I have ended up using 3 listviews and 1 textview for my project, and 2 of those 4 views are essentially "read only".

**Comment by EddRobinson** [comment]
Another thing. So I have my ListView displaying a string list, I'm unsure how to "do stuff" when I click on something in that List.

I can see from the API reference that I need to create an event - OpenSelectedItem.

But I'm not entirely sure how to make it all function. Maybe it's my experience level (2 years in University and currently on an Industrial Year), but I'm having a hard time following a lot of the documentation

**Reply by BDisp** [reply]
What you need to do is subscribe the `OpenSelectedItem` event by setting like this:
```cs
myListView.OpenSelectedItem += (e) => { doSomething };
```
In v2 is:
```cs
myListView.OpenSelectedItem += (s, e) => { doSomething };
```
There is also another event that is `SelectedItemChanged.`

**Reply by EddRobinson** [reply]
Ahhh, many thanks.

**Reply by BDisp** [reply]
You welcome. You also can use a method like this:

```cs
myListView.OpenSelectedItem += DoSomething();

private void DoSomething(object s, ListViewItemEventArgs e)
{
  ...
}

```
